### PR TITLE
Brush uses context instead of cloning

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

No more cloning

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

No cloning!

## How Has This Been Tested?

storybook, npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

**Review by ckifer — APPROVED** [review]
Nice

File: src/context/chartDataContext.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+const ChartDataContext = createContext<any[] | undefined>(undefined);
+const DataStartIndexContext = createContext<number>(0);
+const DataEndIndexContext = createContext<number>(0);
+
+export const ChartDataContextProvider = ChartDataContext.Provider;
+export const DataStartIndexContextProvider = DataStartIndexContext.Provider;
+export const DataEndIndexContextProvider = DataEndIndexContext.Provider;
+
+/**
+ * "data" is the data of the chart - it has no type because this part of recharts is very flexible.
```
[the remark this answers is not in the document]
Comment (PavelVanecek) [reply src/context/chartDataContext.tsx:12]: Different charts have different requirements. Some read `name` and some read `value` and then there's the dataKey feature overriding everything.

I say we leave it as `any[]` for now and look into this for 4.x release.

File: src/context/chartDataContext.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * Basically it's an array of "something" and then there's the dataKey property in various places
+ * that's meant to pull other things away from the data.
+ *
+ * Some charts have `data` defined on the chart root, and they will return the array through this hook.
+ * For example: <ComposedChart data={data} />.
+ *
+ * Other charts, such as Pie, have data defined on individual graphical elements.
+ * These charts will return `undefined` through this hook, and you need to read the data from children.
+ * For example: <PieChart><Pie data={data} />
```
[the remark this answers is not in the document]
Comment (PavelVanecek) [reply src/context/chartDataContext.tsx:21]: I could be wrong.

[the remark this answers is not in the document]
Comment (ckifer) [reply src/context/chartDataContext.tsx:21]: Well from the docs you aren't so all good :)

**Pull request merged**